source/shadow.c

```diff
@@ -196,38 +266,92 @@ static ShadowStatus_t containsSubString( const char * pString,
 
     return returnStatus;
 }
+
 /*-----------------------------------------------------------*/
 
-static ShadowStatus_t validateThingName( const char * pString,
-                                         uint16_t stringLength,
-                                         uint16_t * pThingNameLength )
+static ShadowStatus_t validateName( const char * pString,
+                                    uint16_t stringLength,
+                                    uint16_t * pNameLength )
 {
     uint16_t index = 0U;
-    ShadowStatus_t returnStatus = SHADOW_THINGNAME_PARSE_FAILED;
+    ShadowStatus_t returnStatus = SHADOW_FAIL;
 
     for( ; index < stringLength; index++ )
     {
+        /* The name should always be terminated by a forward slash */
         if( pString[ index ] == ( char ) '/' )
         {
+            if( index > 0U )
+            {
+                /* Only accept names of greater than zero length */
+                *pNameLength = index;
+                returnStatus = SHADOW_SUCCESS;
+            }
+
             break;
         }
     }
 
-    /* Zero length thing name is not valid,
-     * $"$aws/things/<ThingName>/
-     * $"$aws/things/<ThingName>"
-     * will extract the same thing name result.
-     * Only empty thing name string like:
-     * "$aws/things/" or "$aws/things" will fail.
-     */
-    if( index > 0U )
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static ShadowStatus_t extractShadowRootAndName( const char * pTopic,
+                                                uint16_t topicLength,
+                                                uint16_t * pConsumedTopicLength,
+                                                uint16_t * pShadowNameLength )
+{
+    ShadowStatus_t shadowStatus;
+
+    /* Look for the named shadow root */
+    shadowStatus = containsSubString( &( pTopic[ *pConsumedTopicLength ] ),
+                                      topicLength - *pConsumedTopicLength,
+                                      SHADOW_NAMED_ROOT,
+                                      SHADOW_NAMED_ROOT_LENGTH );
+
+    if( shadowStatus == SHADOW_SUCCESS )
     {
-        *pThingNameLength = index;
-        returnStatus = SHADOW_SUCCESS;
+        /* Topic is a named shadow */
+        *pConsumedTopicLength += SHADOW_NAMED_ROOT_LENGTH;
```

Comment:
Does this function has an assumption that the NAMED ROOT substring starts at pTopic[ *pConsumedTopicLength ]. If so it is better to assert or have a validation to confirm that.

Reply:
I'm not sure if I'm missing your point, but just before this is a validation that the NAMED_ROOT exists. So we already know it to be true.

```
    /* Look for the named shadow root */
    shadowStatus = containsSubString( &( pTopic[ *pConsumedTopicLength ] ),
                                      topicLength - *pConsumedTopicLength,
                                      SHADOW_NAMED_ROOT,
                                      SHADOW_NAMED_ROOT_LENGTH );

```

Reply:
How about if the topic string is 
aws/things/<thingName>/**somestring**/shadow/name/<shadowName>/<operation>

Reply:
There are unit test cases for the following four topic strings:

```
#define TEST_TOPIC_STRING_UNTERMINATED_THINGNAME               "$aws/things/TestThingName"
#define TEST_TOPIC_STRING_EMPTY_SHADOW_ROOT                    "$aws/things/TestThingName/"
#define TEST_TOPIC_STRING_INVALID_SHADOW_ROOT                  "$aws/things/TestThingName/invalid"
#define TEST_CLASSIC_TOPIC_STRING_EMPTY_SHADOW_MESSAGE_TYPE    "$aws/things/TestThingName/shadow"
```
The third of those best covers your "somestring" scenario. The error code is SHADOW_ROOT_PARSE_FAILED. extractShadowRootAndName() looks for "/shadow" or "/shadow/name/" (where your "somestring" is located) and returns SHADOW_ROOT_PARSE_FAILED failed if neither is found.

Reply:
Thanks. From the name of the function containsSubString, I was thinking it is invoking strstr internally. Just looked at the implementation and found it to be using strncmp from the start of the first pointer. I think it is better to have a different name for the containsSubString . However, this is not in the scope of this PR. Will address it separately.